### PR TITLE
Added manageiq dynamic resource definition module.

### PR DESCRIPTION
**Pull request by ahmbas**
##### SUMMARY

New module to support create, edit and delete dynamic resource definitions

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

manageiq_dynamic_resource_definition

##### ANSIBLE VERSION
```
ansible 2.5.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

```

### CODE REVIEW

**Review by cben — COMMENTED** [review]
Didn't review in detail yet but looks sane :+1: 
One question: should the module call these "dynamic resource" or "generic object".
AFAICT, api, core models, and UI use the term "generic object":
screenshot
(screenshot stolen from http://talk.manageiq.org/t/solved-dynamic-resource-objects/3122/4)
but [documentation](http://manageiq.org/docs/reference/latest/doc-Provisioning_Virtual_Machines_and_Hosts/miq/index#dynamic-resource-objects) uses "dynamic resource".

EDIT: if the doc is to be trusted, UI does use "Dynamic Resource" in many places?

**Comment by cben** [comment]
Ahem, the doc seems to lie!
There's no Automation > Automate > Dynamic Resource Objects with "Add a New Dynamic Resource Object Class".
I see (on master) Automation > Automate > Generic Objects, shows "Generic Objects Classes" in title & explorer, "Add a new Generic Object Class"...

@cbudz I see you renamed it in docs to "dynamic resource" in https://github.com/ManageIQ/manageiq_docs/pull/553.  Can you shed light on official name, and if it's expected that UI doesn't use the term docs claim it does?

EDIT: also [asked on gitter](https://gitter.im/ManageIQ/manageiq/automate?at=5ad306295f188ccc15681903)

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: this may be what the API returns, but not what this function returns?

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: I think you could make these required=True in module args spec, and here just assume they're set?

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: This tests Python "truthiness".  `False`, `""`, `[]` etc will all fail the check.
Does that match your goals in `edit_definition()`?

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (ahmbas) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: Makes sense, will change it.

Comment (ahmbas) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: Actually after reviewing it, We need to check for properties here, since it is not required if the state is absent but it is required if the state is present. I'll remove the name check since it is required for all states

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (ahmbas) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: If you create a dynamic resource definition via the API and request it again it will return the attributes, associations and methods fields, even if they are not defined. The will be returned as empty.

Here, I just want to check if they exist or not. I will add a TypeError try/except to handle unhashable types.

The end goal is to see if ```['attributes', 'associations', 'methods']``` are keys in the properties dictionary and check if they are empty or not

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: Perhaps `required_if` can help.

Comment (ahmbas) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: :+1:  Didn't know about it. Added it.

**Comment by ahmbas** [comment]
@cben Thanks for the review.
Quick question. I do not understand the ansibot error 
```
lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py:0:0: E324 Value for "default" from the argument_spec ({'verify_ssl': True}) for "manageiq_connection" does not match the documentation (None)
```

I am just reusing the ```manageiq_connection``` module. Could you shed some light on this?

**Comment by cbudz** [comment]
Hey @cben 

I've reached out to get some feedback on how we reference these 'objects' in our docs, as the issue has been raised before and its leading to confusion between what our docs refer to and the actual labeling in the UI. I'll try to update this shortly.

-Chris

**Comment by cben** [comment]
> `lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py:0:0: E324 Value for "default" from the argument_spec ({'verify_ssl': True}) for "manageiq_connection" does not match the documentation (None)`

@ahmbas dunno, looks like linter bug.  You do have `extends_documentation_fragment: manageiq` which includes doc  https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/manageiq.py#L31 which documents `verify_ssl` default `true`.

Ignore this for now, for a new module you'll need a core Ansible reviewer anyway and I suppose they'd know.

Reviewing more...

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: please document what this consists of.  you may use "suboptions:" (see manageiq_provider for example), or just prose.

hmm, do you need this nesting at all?
how about this module taking `attributes:` and `associations:` args at top level?

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: I think this validation can be dropped too.  it's only `name` which is required=True.
(plus we wouldn't even reach this function without finding a definition by name)

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: OK, I see what you're doing here :+1:, I think there is slightly better way.  Just suggestions, your call.

Just removing these is not perfect.  Ideally, omitting say `associations` in module args would behave same as writing explicit `associations: {}`.  But currently explicit `associations: {}` won't match what's left after you remove empty values from API, causing unnecessary request and changed=True.
- Args side can be normalized for you if you make attributes & associations first-class options, or suboptions, defaulting to `{}`.
  - Note that only recently manageiq master started sending empty values; nulls were omitted before.  Not sure that's relevant — do you get Nones or empty hashes here?  Anyway :+1: to normalizing API results in one direction or another — don't *rely* on getting empty values.
- A pattern that works & reads nicely is defining a normalizing function and applying it on both sides of comparison.  `if normalize(current) == normalized(desired)`.
- IMHO, splitting logic between `has_field` and here is harder to follow, and it could be simple enough to inline, perhaps something like this:
  ```
  current_properties = dynamic_resource_definition.get('properties', {})
  current_name = dynamic_resource_definition.get('name')
  if (current_name == name and
      current_properties.get('attributes', {}) == properties.get('attributes', {}) and
      current_properties.get('associations', {}) == properties.get('associations', {}):
  ```

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (ahmbas) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: Sure, Will remove it.

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (ahmbas) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: I see what you mean. I agree, the inline solution is more simple and straight forward

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: ooh, try removing above 2 lines and see if it makes CI happier.
 `manageiq_connection` suboptions (including `verify_ssl`) documentation should be inherited via `extends_documentation_fragment: manageiq`, and having this here might interfere with that...

**Comment by cben** [comment]
Looks great :+1:
shipit

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: also remove this line.  you already have it at top level on L21.  
this is what causes the  `'AnsibleUnicode' object has no attribute 'pop'` errors.  here under `options:` ansible-doc assumes all entries will have a sub-hash as value...
(ansible-doc is unhelpful when it crashes, but I run `bin/ansible-doc manageiq_dynamic_resource_definition -vvv` to get traceback, and added print statements from there...)

**Comment by cben** [comment]
OK, seems the remaining error is common.
The check was added in #36247, see list of errors https://gist.github.com/sivel/47ffa37450ec3512551fe9d561c25a4c — lots of "E324 ... (True) ...  does not match the documentation (None)".

=> simply add your new module to `test/sanity/validate-modules/ignore.txt`.  all other manageiq modules are already there for E324.

**Comment by ahmbas** [comment]
@cben Perfect. I'll add the ignore line

File: test/sanity/validate-modules/ignore.txt
Comment (mattclay) [inline test/sanity/validate-modules/ignore.txt]: New ignores should not be added. The reported issues should be fixed. If the issues exist in common docs fragments then those should be fixed in a separate PR first.

**Review by mattclay — DISMISSED** [review]

File: test/sanity/validate-modules/ignore.txt
Comment (cben) [reply test/sanity/validate-modules/ignore.txt]: @mattclay see [above](https://github.com/ansible/ansible/pull/38725#issuecomment-381643640) why I believe linter is wrong here.  Common doc fragment & argspec agree on default here.

`test/sanity/validate-modules/main.py` does understand shared doc fragments.
But it gets confused by suboptions that have a default:
https://gist.github.com/cben/14de325ceec502d86c1edbcd082be9c2
There is `TODO: needs to recursively traverse suboptions` :-).  I'm not familiar with it enough really implement that, but I'm working on patch to not produce spurious error in such scenario.

Comment (cben) [reply test/sanity/validate-modules/ignore.txt]: WIP #38962.

Comment (sivel) [reply test/sanity/validate-modules/ignore.txt]: This error has brought to my attention a bug in the options parser.  The use of `default` on an option using `options`  with `default` on the suboptions, is a work around for the real problem.

A PR with the resolution can be found at: https://github.com/ansible/ansible/pull/38967

Once this is resolved, `default=` can be removed from the `manageiq_connection` option.

Comment (cben) [reply test/sanity/validate-modules/ignore.txt]: Great find!  I forgot all about the workround in manageiq modules (added in #31774) despite reviewing it :)

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (gtanzillo) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: I think this is a typo - `changed` should be `changes`

**Review by gtanzillo — APPROVED** [review]
LGTM 👍 @abelllotti can you give this a once over and make sure it looks ok from the ManageIQ API perspective

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (abellotti) [inline lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py:140]: may be nice to support description in addition to name.

File: lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py
Comment (cben) [reply lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_definition.py]: no, this refers to the standard `changed=True` / `changed=False` and `msg="..."` results returned by ansible modules.

**Comment by cben** [comment]
#38967 landed, I expect CI will now pass.
shipit
stale_ci
(if this doesnt rerun CI, try closing/reopening)

**Comment by yaacov** [comment]
shipit

**Comment by ahmbas** [comment]
@cben Quick question. I just rebased my branch to keep it up to date with devel.
Now the shipit label is gone and the bot added the community review label again.
Is there a way around this?

**Comment by cben** [comment]
Right, the way is we repeat it :-)
shipit

**Comment by yaacov** [comment]
shipit

**Comment by cben** [comment]
@sivel @mattclay I think this needs a core reviewer (new module)?

**Comment by ahmbas** [comment]
@cben @sivel @mattclay any update on this?

**Comment by mattclay** [comment]
@ahmbas `version_added` should be 2.7:

https://github.com/ansible/ansible/pull/38725/files#diff-bd2070f489176c470af4ef8da01bff7eR21

**Comment by ahmbas** [comment]
@mattclay Done

**Pull request closed without merging**
